adapter/prometheus2/prometheus.go

```diff
@@ -0,0 +1,287 @@
+		registry *prometheus.Registry
+		once     sync.Once
+	}
+
+	handler struct {
+		metrics map[string]prometheus.Collector
+		kinds   map[string]config.Params_MetricInfo_Kind
+	}
+)
+
+var (
+	charReplacer = strings.NewReplacer("/", "_", ".", "_", " ", "_")
+
+	_ metric.HandlerBuilder = &builder{}
+	_ metric.Handler        = &handler{}
+)
+
+// GetBuilderInfo returns the BuilderInfo associated with this adapter implementation.
+func GetBuilderInfo() adapter.BuilderInfo {
+	return adapter.BuilderInfo{
+		Name:        "prometheus",
+		Description: "Publishes prometheus metrics",
+		SupportedTemplates: []string{
+			metric.TemplateName,
+		},
+		CreateHandlerBuilder: func() adapter.HandlerBuilder { return newFactory(newServer(defaultAddr)) },
+		DefaultConfig:        &config.Params{},
+		ValidateConfig:       func(msg adapter.Config) *adapter.ConfigErrors { return nil },
+	}
+}
+
+func newFactory(srv server) *builder {
+	return &builder{srv, prometheus.NewPedanticRegistry(), sync.Once{}}
+}
+
+func (b *builder) ConfigureMetricHandler(map[string]*metric.Type) error { return nil }
+
+func (b *builder) Build(c adapter.Config, env adapter.Env) (adapter.Handler, error) {
+	var serverErr error
+	b.once.Do(func() {
+		serverErr = b.srv.Start(env, promhttp.HandlerFor(b.registry, promhttp.HandlerOpts{}))
+	})
+	if serverErr != nil {
+		return nil, fmt.Errorf("could not start prometheus server: %v", serverErr)
+	}
+
+	cfg := c.(*config.Params)
+	var metricErr *multierror.Error
+	metricsMap := make(map[string]prometheus.Collector, len(cfg.Metrics))
+	kinds := make(map[string]config.Params_MetricInfo_Kind, len(cfg.Metrics))
+	for _, m := range cfg.Metrics {
+		switch m.Kind {
+		case config.GAUGE:
+			c, err := registerOrGet(b.registry, newGaugeVec(m.Name, m.Description, m.LabelNames))
+			if err != nil {
+				metricErr = multierror.Append(metricErr, fmt.Errorf("could not register metric: %v", err))
+				continue
+			}
+			metricsMap[m.Name] = c
+			kinds[m.Name] = config.GAUGE
+		case config.COUNTER:
+			c, err := registerOrGet(b.registry, newCounterVec(m.Name, m.Description, m.LabelNames))
+			if err != nil {
+				metricErr = multierror.Append(metricErr, fmt.Errorf("could not register metric: %v", err))
+				continue
+			}
+			metricsMap[m.Name] = c
+			kinds[m.Name] = config.COUNTER
+		case config.DISTRIBUTION:
+			c, err := registerOrGet(b.registry, newHistogramVec(m.Name, m.Description, m.LabelNames, m.Buckets))
+			if err != nil {
+				metricErr = multierror.Append(metricErr, fmt.Errorf("could not register metric: %v", err))
+				continue
+			}
+			metricsMap[m.Name] = c
+			kinds[m.Name] = config.DISTRIBUTION
+		default:
+			metricErr = multierror.Append(metricErr, fmt.Errorf("unknown metric kind (%d); could not register metric", m.Kind))
+		}
+	}
+	return &handler{metricsMap, kinds}, metricErr.ErrorOrNil()
+}
+
+// TODO: figure out ownership semantics, because builder.close will never be called in the new scheme.
+func (b *builder) Close() error {
```

Comment:
I think that tells you you're doing something wrong ;-)

The server you're creating in the Build method should be associated with the resulting handler, not with the builder. The builder's job is mostly to accumulate configuration state and then produce a handler. That's it, it shouldn't have any need for heavier lifetime semantics.